Content.Client/Power/PowerMonitoringWindow.xaml.cs

```diff
@@ -1,84 +1,219 @@
+            !MathHelper.CloseToPercent(totalSources + totalBatteryUsage, totalLoads, 0.1f) ? new Color(180, 0, 0) : Color.White;
+
+        // Update lists
+        var generatorList = allEntries.Where(x => x.Group == PowerMonitoringConsoleGroup.Generator);
+        UpdateAllConsoleEntries(SourcesList, generatorList.ToArray(), null, focusLoads);
+
+        var smesList = allEntries.Where(x => x.Group == PowerMonitoringConsoleGroup.SMES);
+        UpdateAllConsoleEntries(SMESList, smesList.ToArray(), focusSources, focusLoads);
+
+        var substationList = allEntries.Where(x => x.Group == PowerMonitoringConsoleGroup.Substation);
+        UpdateAllConsoleEntries(SubstationList, substationList.ToArray(), focusSources, focusLoads);
+
+        var apcList = allEntries.Where(x => x.Group == PowerMonitoringConsoleGroup.APC);
+        UpdateAllConsoleEntries(ApcList, apcList.ToArray(), focusSources, null);
+
+        // Update system warnings
+        UpdateWarningLabel(flags);
+    }
+
+    private void AddTrackedEntityToNavMap(EntityUid uid, EntityCoordinates coords, PowerMonitoringConsoleEntry entry, bool useDarkColors = false)
+    {
+        if (!NavMap.Visible)
+            return;
+
+        var colorMap = useDarkColors ? PowerMonitoringHelper.DarkPowerIconColors : PowerMonitoringHelper.PowerIconColors;
+        var color = uid == _trackedEntity ? Color.White : colorMap[entry.Group];
+
+        var iconPath = PowerMonitoringHelper.CircleIconPath;
+
+        switch (entry.Group)
         {
-            var ent = listVal[i];
-            _prototypeManager.TryIndex(ent.IconEntityPrototypeId, out EntityPrototype? entityPrototype);
-            IRsiStateLike? iconState = null;
-            if (entityPrototype != null)
-                iconState = _spriteSystem.GetPrototypeIcon(entityPrototype);
-            var icon = iconState?.GetFrame(RsiDirection.South, 0);
-            var item = list[i];
-            item.Text = $"{ent.NameLocalized} {Loc.GetString("power-monitoring-window-value", ("value", ent.Size))}";
-            item.Icon = icon;
+            case PowerMonitoringConsoleGroup.Generator:
+                iconPath = PowerMonitoringHelper.CircleIconPath; break;
+            case PowerMonitoringConsoleGroup.SMES:
+                iconPath = PowerMonitoringHelper.HexagonIconPath; break;
+            case PowerMonitoringConsoleGroup.Substation:
+                iconPath = PowerMonitoringHelper.SquareIconPath; break;
+            case PowerMonitoringConsoleGroup.APC:
+                iconPath = PowerMonitoringHelper.TriangleIconPath; break;
         }
+
+        var icon = _spriteSystem.Frame0(new SpriteSpecifier.Texture(new(iconPath)));
+
+        // We expect a single tracked entity at a given coordinate
+        if (NavMap.TrackedEntities.ContainsKey(coords))
+            NavMap.TrackedEntities[coords] = (true, color, icon);
+
+        else
+            NavMap.TrackedEntities.Add(coords, (true, color, icon));
```

Comment:
ContainsKey is redundant when you set it to the same thing anyway.

Reply:
Done